test/tests/functional/pbs_checkpoint.py

```diff
@@ -0,0 +1,88 @@
+# Software Foundation, either version 3 of the License, or (at your option) any
+# later version.
+#
+# PBS Pro is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.
+# See the GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Commercial License Information:
+#
+# For a copy of the commercial license terms and conditions,
+# go to: (http://www.pbspro.com/UserArea/agreement.html)
+# or contact the Altair Legal Department.
+#
+# Altair’s dual-license business model allows companies, individuals, and
+# organizations to create proprietary derivative works of PBS Pro and
+# distribute them - whether embedded or bundled with other software -
+# under a commercial license agreement.
+#
+# Use of Altair’s trademarks, including but not limited to "PBS™",
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+
+
+from tests.functional import *
+
+
+class TestCheckpoint(TestFunctional):
+
+    """
+    This test suite targets Checkpoint functionality.
+    """
+    def test_checkpoint_abort_with_preemept(self):
+        abort_script = """#!/bin/bash
+kill $1
+exit 0
+"""
+        pbs_home = self.server.pbs_conf['PBS_HOME']
+        abort_file = os.path.join(pbs_home, 'mom_priv', 'abort_file')
+        localhost = socket.gethostname()
+        fn = self.du.create_temp_file(hostname=localhost, body=abort_script)
+        ret = self.du.run_copy(src=fn,
+                               dest=abort_file, sudo=True, uid='root',
+                               gid='root', mode=0755)
+        if ret['rc'] != 0:
+            self.skipTest('failed to create abort script')
+        c = {'$action': 'checkpoint_abort 30 !' + abort_file + ' %sid'}
+        self.mom.add_config(c)
+        self.scheduler.set_sched_config({'preempt_order': 'C'})
+        a = {'queue_type': 'e',
+             'started': 'True',
+             'enabled': 'True',
+             'Priority': 150}
+        self.server.manager(MGR_CMD_CREATE, QUEUE, a, "expressq")
+        node_attrs = self.server.status(NODE, id=self.server.shortname)[0]
+        ncpus = node_attrs['resources_available.ncpus']
+        attrs = {'Resource_List.select': '1:ncpus=' + ncpus}
+        j1 = Job(TEST_USER, attrs)
+        j1.set_sleep_time(5)
+        jid1 = self.server.submit(j1)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid1)
+        attrs['queue'] = 'expressq'
+        j2 = Job(TEST_USER, attrs)
+        j2.set_sleep_time(1)
+        self.server.submit(j2)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid1)
+        time.sleep(30)
+        self.server.expect(JOB, {'job_state': 'H'}, id=jid1)
```

Comment:
@bayucan I think the job should ideally not go into the held state because it should just run as soon as the high priority job ends. The reason it is happening that way is because of another bug in PBS. Once that bug is resolved, this test will start failing.
I had a chat about this test with Bhroam and we think we can make this a generic checkpointing test that checks the first job ran and finished after the high priority job ends and also check for the node to be in free state (which you are already doing).
This will be a generic test and can also go into our SmokeTest test suite.

Reply:
Yes, there's a checkpoint abort bug right now that is preventing the first job into going into 'S' state but instead, goes to 'E' state and eventually to 'H' state. Once the job is in 'H' state and I issue a qdel-ed, then I get a mom crash, which is the situation this test is trying to capture. For now, I'll add a comment to mention all of this, and that once the checkpoint abort bug is fixed, we can check for the job going into 'S' state. I've also added code to check that the high priority job ran and finished so it can also look like a generic checkpoint test.